### PR TITLE
Reload new config without restarting process

### PR DESCRIPTION
**Pull request by ganmacs**
**Which issue(s) this PR fixes**: 
Fixes https://github.com/fluent/fluentd/issues/2624

**What this PR does / why we need it**: 

This change make fluentd be able to reload new config without restarting process with SIGUSR2 signal. 

it's very lighter and safer than existing reloading feature. but it has 2 limitations.

1.  A change to system_config is ignored because it needs to restart(kill/spawn) process. 
2. All plugins must not use class variable when restarting.

**Docs Changes**:

Add USR2 description.

**Release Note**: 

same as title

### CODE REVIEW

File: lib/fluent/engine.rb [hunk 1]
```diff
@@ -157,14 +158,54 @@ def run
...
+    # @param supervisor [Bool]
+    # @reutrn nil
+    def reload_config(conf, supervisor: false)
+      # configure first to reduce down time while restarting
+      new_agent = RootAgent.new(log: log, system_config: @system_config)
+      ret = Fluent::StaticConfigAnalysis.call(conf, workers: system_config.workers)
+
+      ret.all_plugins.each do |plugin|
+        if plugin.respond_to?(:reloadable_plugin?) && !plugin.reloadable_plugin?
+          raise Fluent::ConfigError, "Unreloadable plugin: #{plugin.class}"
```
Comment (cosmo0920) [inline lib/fluent/engine.rb]: This code seems good but users doesn't always concern about their using plugin implemented class.
Can we output `plugin_name` in Fluentd configuration(`@type plugin_name`)?

File: lib/fluent/engine.rb [hunk 2]
```diff
@@ -157,14 +158,54 @@ def run
...
+      end
+
+      stop_phase(old_agent)
+
+      # Restart phase
+      new_fleunt_log_event_router = FluentLogEventRouter.build(new_agent)
+      if new_fleunt_log_event_router.emittable?
+        $log.enable_event(true)
+      end
+      @fluent_log_event_router = new_fleunt_log_event_router
```
Comment (cosmo0920) [inline lib/fluent/engine.rb]: nits: Why not extract here as `restart_phase` method?
I'm slightly a bit confused here due to method implementation granularity glitch.

File: lib/fluent/engine.rb (hunk 1 above)
Comment (ganmacs) [reply lib/fluent/engine.rb]: fixed https://github.com/fluent/fluentd/pull/2716/commits/a78f7de7f4791faf97010be9c378f67d772b27b5

File: lib/fluent/engine.rb (hunk 2 above)
Comment (ganmacs) [reply lib/fluent/engine.rb]: fixed https://github.com/fluent/fluentd/pull/2716/commits/37f64c6ff2b4c3d0d87107cc7e59a43bc87daba1

**Comment by ganmacs** [comment]
@repeatedly ?

File: lib/fluent/static_config_analysis.rb [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+require 'fluent/config'
+require 'fluent/plugin'
+
+module Fluent
+  # Static Analysis means analysing all plugins and Fluent::Element without invokeing Plugin#configure
+  class StaticConfigAnalysis
+    module Elem
+      Input = Struct.new(:plugin, :config)
+      Output = Struct.new(:plugin, :config)
+      Format = Struct.new(:plugin, :config)
```
Comment (repeatedly) [inline lib/fluent/static_config_analysis.rb]: Why `Format`, not `Filter`?

File: test/test_config.rb [hunk 4]
```diff
@@ -175,5 +175,27 @@ def create_warn_dummy_logger
     logger = ServerEngine::DaemonLogger.new(logdev, dl_opts)
     $log = Fluent::Log.new(logger)
   end
-end
 
+  sub_test_case '.build'do
```
Comment (repeatedly) [inline test/test_config.rb]: Need space before `do`

File: lib/fluent/plugin/buf_file.rb [hunk 5]
```diff
@@ -43,19 +44,20 @@ class FileBuffer < Fluent::Plugin::Buffer
...
         @multi_workers_available = false
         @additional_resume_path = nil
         @buffer_path = nil
+        @variable_store = nil
       end
 
       def configure(conf)
         super
 
+        @variable_store = Fluent::VariableStore.fetch_or_build(:buf_file_single)
```
[the remark this answers is not in the document]
Comment (ganmacs) [reply lib/fluent/plugin/buf_file.rb]: fixed https://github.com/fluent/fluentd/pull/2716/commits/500b0254587b85a44c8af932642b7b094fd3a56d

File: test/test_config.rb (hunk 4 above)
Comment (ganmacs) [reply test/test_config.rb]: fixed https://github.com/fluent/fluentd/pull/2716/commits/500b0254587b85a44c8af932642b7b094fd3a56d

File: lib/fluent/static_config_analysis.rb (hunk 3 above)
Comment (ganmacs) [reply lib/fluent/static_config_analysis.rb]: Good catch... https://github.com/fluent/fluentd/pull/2716/commits/60440d5fe931e3964b4bf9c05bb8ea8a70cfee75

**Comment by repeatedly** [comment]
I noticed we need to change HTTP RPC for reload.
Should we change this to use USR2?

https://github.com/fluent/fluentd/blob/c8097882f6485e15da8b396c18f1b65a97c570e7/lib/fluent/supervisor.rb#L106

**Comment by cosmo0920** [comment]
> Should we change this to use USR2 or add new path?

I believe that we should add a new path for reloading config.
Windows does not have `USR2` signal:
* https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/signal?view=vs-2019

**Comment by ganmacs** [comment]
> I noticed we need to update HTTP RPC for this feature.
Should we change this to use USR2 or add new path?

Good catch! I forgot about it. I added new endpoint `/api/config.gracefulReload`(https://github.com/fluent/fluentd/pull/2716/commits/d8b8bc6ad4aec438eba96a695a59266be03bcc34) for compatibility.

> Windows does not have USR2 signal:

I think this feature can't support windows like `USR1`  because a supervisor needs to send `USR2` signal to workers to do this. https://github.com/fluent/fluentd/pull/2716/files#diff-dbe0e1ec4079138e48ca6a4d7c7248f9R215
We need to introduce another way to communicate between supervisor and workers without using a signal to be able to use this feature in windows.

**Comment by repeatedly** [comment]
I will release v1.9.1.rc1 soon.

**Pull request merged**